Source/Core/Core/PatchEngine.cpp

```diff
@@ -203,30 +204,63 @@ static void ApplyPatches(const std::vector<Patch>& patches)
   }
 }
 
-void ApplyFramePatches()
+// Requires MSR.DR, MSR.IR
+// There's no perfect way to do this, it's just a heuristic.
+// We require at least 2 stack frames, if the stack is shallower than that then it won't work.
+static bool IsStackSane()
 {
-  // TODO: Messing with MSR this way is really, really, evil; we should
-  // probably be using some sort of Gecko OS-style hooking mechanism
-  // so the emulated CPU is in a predictable state when we process cheats.
-  u32 oldMSR = MSR;
-  UReg_MSR newMSR = oldMSR;
-  newMSR.IR = 1;
-  newMSR.DR = 1;
-  MSR = newMSR.Hex;
+  _dbg_assert_(ACTIONREPLAY, UReg_MSR(MSR).DR && UReg_MSR(MSR).IR);
+
+  // Check the stack pointer
+  u32 SP = GPR(1);
+  if (!PowerPC::HostIsRAMAddress(SP))
+    return false;
+
+  // Read the frame pointer from the stack (find 2nd frame from top), assert that it makes sense
+  u32 next_SP = PowerPC::HostRead_U32(SP);
+  if (next_SP <= SP || !PowerPC::HostIsRAMAddress(next_SP) ||
+      !PowerPC::HostIsRAMAddress(next_SP + 4))
+    return false;
+
+  // Check the link register makes sense (that it points to a valid IBAT address)
+  auto insn = PowerPC::TryReadInstruction(PowerPC::HostRead_U32(next_SP + 4));
```

Comment:
`Read_Opcode` is what generates ISIs AFAICT. `TryReadInstruction` modifies the instruction cache which was not intended.

I only used `TryReadInstruction` because it seemed to be the only API that did approximately what I needed (validate if an address could be successfully translated to a valid physical address through the IBAT) but it has too many caveats so it wasn't correct to use it.